### ISSUE TITLE
PDF output mangling image for TIFF input

### ISSUE THREAD

**Issue by jbreiden** [tesseract-ocr/tesseract#535]
This means api->GetInputImage() is giving us a processed image.

[test.tif.zip](https://github.com/tesseract-ocr/tesseract/files/632436/test.tif.zip)
[test.pdf](https://github.com/tesseract-ocr/tesseract/files/632437/test.pdf)

**Comment by jbreiden** [comment]
Emergency workaround while I go hunt down root cause.

```
--- tesseract/api/pdfrenderer.cpp	2016-11-21 08:45:47.000000000 -0800
+++ tesseract/api/pdfrenderer.cpp	2016-12-05 14:15:42.000000000 -0800
@@ -841,8 +841,8 @@
 bool TessPDFRenderer::AddImageHandler(TessBaseAPI* api) {
   size_t n;
   char buf[kBasicBufSize];
-  Pix *pix = api->GetInputImage();
   char *filename = (char *)api->GetInputName();
+  Pix *pix = pixRead(filename);
   int ppi = api->GetSourceYResolution();
   if (!pix || ppi <= 0)
     return false;
```

**Comment by jbreiden** [comment]
I think this is where we are damaging our input image

https://github.com/tesseract-ocr/tesseract/blob/master/api/baseapi.cpp#L692

**Comment by jbreiden** [comment]
This one is probably best.

```c++
--- tesseract/ccmain/thresholder.cpp	2016-03-11 14:29:36.000000000 -0800
+++ tesseract/ccmain/thresholder.cpp	2016-12-05 15:00:46.000000000 -0800
@@ -225,7 +225,7 @@
 Pix* ImageThresholder::GetPixRect() {
   if (IsFullImage()) {
     // Just clone the whole thing.
-    return pixClone(pix_);
+    return pixCopy(pix_);
   } else {
     // Crop to the given rectangle.
     Box* box = boxCreate(rect_left_, rect_top_, rect_width_, rect_height_);
@@ -322,4 +322,3 @@
 }
 
 }  // namespace tesseract.
-
```

**Comment by jbreiden** [comment]
This bug happens when:

  * input image is binary, which causes up to corrupt api->GetInputImage()
  * and not (JPEG2000 || JPEG ||  PNG), which causes us to utilize api->GetInputImage()

So for example, this example is TIFF G4. Converting to an identical looking TIFF LZW 
grayscale does not tickle this bug.

**Comment by jbreiden** [comment]
Ray found the exact spot. This is the final answer.

```c++
--- tesseract/ccmain/thresholder.cpp	2016-03-11 14:29:36.000000000 -0800
+++ tesseract/ccmain/thresholder.cpp	2016-12-05 15:27:45.000000000 -0800
@@ -181,8 +181,9 @@
 // Caller must use pixDestroy to free the created Pix.
 void ImageThresholder::ThresholdToPix(PageSegMode pageseg_mode, Pix** pix) {
   if (pix_channels_ == 0) {
-    // We have a binary image, so it just has to be cloned.
-    *pix = GetPixRect();
+    // We have a binary image, so it just has to be copied.
+    // Don't clone or you'll mess up api->GetInputImage()
+    *pix = pixCopy(NULL, GetPixRect());
   } else {
     OtsuThresholdRectToPix(pix_, pix);
   }
@@ -322,4 +323,3 @@
 }
 
 }  // namespace tesseract.
-
```

**Comment by jbreiden** [comment]
Note that this bug affects all versions of Tesseract capable of producing PDF output, bot 3.0.x and 4.x.

**Comment by jbreiden** [comment]
... And the code above is leaky. Ray is doing the final final final version right now.

**Comment by theraysmith** [comment]
Fixed in commit 7744da9..025689f.
